### PR TITLE
Filter root spans

### PR DESCRIPTION
**Pull request by stoeffel**
filtering

### CODE REVIEW

File: nri-log-explorer/package.yaml [hunk 1]
```diff
@@ -22,11 +22,14 @@ executables:
     - brick >= 0.51 && < 1.0
     - bytestring >= 0.10.8.2 && < 0.12
     - directory >= 1.3.3.0 && < 1.4
+    - fuzzy >= 0.1.0.0 && < 0.2
```
Comment (jwoudenberg-nri) [inline nri-log-explorer/package.yaml:25]: 💭 This seems to pull in a big dependency `monoid-subclasses`. Might be worth seeing if we can drop it at some point.

File: nri-log-explorer/src/Main.hs [hunk 2]
```diff
@@ -42,15 +45,18 @@ import qualified System.Exit
...
+import qualified Text.Fuzzy as Fuzzy
 import qualified Prelude
 
 data Model = Model
   { -- Used to calculate "2 minutes ago" type info for root spans.
     currentTime :: Time.UTCTime,
     -- A tool like pbcopy or xclip for copying to clipboard.
     clipboardCommand :: Maybe Text,
+    -- All root spans.
+    allRootSpans :: List RootSpan,
```
Comment (jwoudenberg-nri) [inline nri-log-explorer/src/Main.hs:57]: ❔ Given that `ListWidget.List` is backed by a `Vector`, I wonder if keeping the original list in a `Vector` as well will mean we need to do fewer conversions when updating a filter.

File: nri-log-explorer/src/Main.hs [hunk 3]
```diff
@@ -59,9 +65,15 @@ data Model = Model
...
     -- hands off the controls.
-    userDidSomething :: Bool
+    userDidSomething :: Bool,
+    filter :: Filter
   }
 
+data Filter
+  = NoFilter
+  | HasFilter Text (List Text)
+  | Filtering (Maybe (Text, List Text)) (Edit.Editor Text Name)
```
Comment (jwoudenberg-nri) [inline nri-log-explorer/src/Main.hs:75]: 😻

Comment (jwoudenberg-nri) [reply nri-log-explorer/src/Main.hs:75]: 🎨 Maybe rename `Filtering` to `EditingFilter`?

File: nri-log-explorer/src/Main.hs [hunk 4]
```diff
@@ -82,6 +94,11 @@ data Msg
   | Exit
   | SetCurrentTime Time.UTCTime
   | CopyDetails
+  | ShowFilter
+  | ClearFilter
+  | StopFiltering (Maybe (Text, List Text))
```
Comment (jwoudenberg-nri) [inline nri-log-explorer/src/Main.hs:99]: 🎨 Maybe `StopEditingFilter`?

File: nri-log-explorer/src/Main.hs [hunk 5]
```diff
@@ -90,34 +107,35 @@ data Msg
...
+  | FilterField
   deriving (Eq, Ord, Show)
 
 -- An alternative data type containing part of the same data as above, in a
 -- format more convenient for some update and view functions.
 data Page
-  = NoDataPage
-  | RootSpanPage Time.UTCTime (ListWidget.List Name RootSpan)
+  = NoDataPage Filter
+  | RootSpanPage Time.UTCTime Filter (ListWidget.List Name RootSpan)
```
Comment (jwoudenberg-nri) [inline nri-log-explorer/src/Main.hs:117]: 😻

File: nri-log-explorer/src/Main.hs (hunk 2 above)
Comment (stoeffel) [reply nri-log-explorer/src/Main.hs]: done

**Comment by stoeffel** [comment]
@jwoudenberg-nri I've addressed all your comments.
I think there are no alternative packages for fuzzy that I can find on hackage :-(
I think our options are drop fuzzy filtering (it's so nice though) or write/fork our own fuzzy. I'm happy to fork, but I think that's out of scope for this PR, except if you think it's blocking it. Thoughts?

**Review by jwoudenberg-nri — APPROVED** [review]
This is great, thank you so much!

**Comment by jwoudenberg-nri** [comment]
> @jwoudenberg-nri I've addressed all your comments.

Sorry, was still reviewing so I added a few more!

> I think there are no alternative packages for fuzzy that I can find on hackage :-(
> I think our options are drop fuzzy filtering (it's so nice though) or write/fork our own fuzzy. I'm happy to fork, but I think that's out of scope for this PR, except if you think it's blocking it. Thoughts?

I definitely want to keep fuzzy searching. I wonder if creating our own function for matching would be a lot of work, but definitely agree it's out of scope for this PR. Just a thought that occurred to me while reviewing, nothing I feel we need to act on now!

File: nri-log-explorer/src/Main.hs [hunk 6]
```diff
@@ -198,6 +215,75 @@ update model msg =
...
+      let editor = Edit.editorText FilterField (Just 1) ""
+       in continueAfterUserInteraction
+            model
+              { filter =
+                  case filter model of
+                    NoFilter -> EditFilter Nothing editor
+                    EditFilter previous _ -> EditFilter previous editor
+                    HasFilter first rest ->
+                      EditFilter (Just (first, rest))
+                        <| Edit.applyEdit (\_ -> TZ.gotoEOL <| TZ.textZipper [Text.join " " (first : rest)] (Just 1)) editor
```
[the remark this answers is not in the document]
Comment (stoeffel) [reply nri-log-explorer/src/Main.hs:229]: We do get live updates as we type.
Might make it a little simpler though. will check

File: nri-log-explorer/src/Main.hs [hunk 7]
```diff
@@ -530,36 +698,55 @@ handleEvent ::
...
-        Vty.EvKey (Vty.KChar 'y') [] -> do
-          liftIO (pushMsg CopyDetails)
-          Brick.continue model
-        -- Fallback
-        _ ->
-          scroll
-            (ListWidget.handleListEventVi ListWidget.handleListEvent vtyEvent)
-            model
-            |> andThen Brick.continue
+      case filter model of
```
[the remark this answers is not in the document]
Comment (stoeffel) [reply nri-log-explorer/src/Main.hs:701]: I think we can't call `handleEditorEvent` in "update" though right?

File: nri-log-explorer/src/Main.hs (hunk 6 above)
[the remark this answers is not in the document]
Comment (stoeffel) [reply nri-log-explorer/src/Main.hs:229]: Actually, I think it makes it harder. Because we also want to track previous filters for the esc behaviour.

**Pull request merged**